sbi/simulators/simutils.py

```diff
@@ -1,83 +1,92 @@
 
-    Distributes all parameters theta to simulate_in_batches, i.e. splits them between
-    cores and then processes them in batches.
+    Parameters are batched with size `sim_batch_size` (default whole theta at once).
 
     Args:
-        simulator: Simulator function.
+        simulator: Simulator callable (a function or a class with `__call__`).
         theta: Parameters $\theta$ sampled from prior or posterior.
-        number_of_workers: Number of parallel workers to start.
-        worker_batch_size: Number of parameters each worker handles. This number of
-            thetas will then be handled by `simulate_in_batches()`.
         sim_batch_size: Number of simulations per batch. Default is to simulate
             the entire theta in a single batch.
-        show_progressbar: Whether to show a progressbar during simulating.
-        logging_level: Minimum severity of messages to log. One of the strings
-            "info", "warning", "debug", "error" and "critical".
-
-    Returns: Parameters theta and simulation outputs x.
+        show_progressbar: Whether to show a progressbar during simulation.
 
+    Returns:
+        Parameters theta and simulations $x$.
     """
 
-    assert worker_batch_size >= sim_batch_size, (
-        "worker_batch_size has to be larger than simulation_batch_size when using"
-        " multiprocessing."
-    )
-
-    batched_simulator = lambda theta_: simulate_in_batches(
-        simulator, theta_, sim_batch_size, show_progressbar=False
-    )
-    theta, x = simulate_mp(
-        simulator=batched_simulator,
-        theta=theta,
-        num_workers=number_of_workers,
-        worker_batch_size=worker_batch_size,
-        show_progressbar=show_progressbar,
-        logging_level=logging_level,
-    )
+    num_sims, *_ = theta.shape
+
+    if num_sims == 0:
+        logging.warning("Zero-length parameter theta implies zero simulations.")
+        x = torch.tensor([])
+    elif sim_batch_size is not None and sim_batch_size < num_sims:
+        # Dev note: pyright complains of torch.split lacking a type stub
+        # as of PyTorch 1.4.0, see https://github.com/microsoft/pyright/issues/291
+        batches = torch.split(theta, sim_batch_size, dim=0)
+
+        pbar = tqdm(
+            total=num_sims,
+            disable=not show_progressbar,
+            desc=f"Running {num_sims} simulations.",
+        )
+
+        with pbar:
+            simulation_outputs = []
+            for batch in batches:
+                simulation_outputs.append(simulator(batch))
+                pbar.update(sim_batch_size)
+
+        x = torch.cat(simulation_outputs, dim=0)
+    else:
+        x = simulator(theta)
 
     return theta, x
 
 
-def simulate_in_batches(
+def simulate_in_batches_joblib(
     simulator: Callable,
     theta: Tensor,
     sim_batch_size: Optional[int] = None,
+    num_workers: int = 1,
     show_progressbar: bool = True,
+    pbar_steps: int = 10,
```

Comment:
The user can not set this argument from outside, no?

Reply:
No, at the moment not. I wanted to avoid yet another kwarg. Do you think we should add it? Maybe we can converge on a visually appealing number of steps in the pbar, I thought `10` would do it.

Reply:
I'd prefer a kwarg. Especially because we will have the simple/advanced interface anyway. So for the simple interface we can take it out.

Reply:
I agree to limit the complexity of the API. There should be a number that works, it boils down to precision in reporting of progress and maybe 10% or whatever it ends up being (5%? 3%?) is enough?

Reply:
If we keep 10 then it will be (10 * `simulation_batch_size`) % steps.

Reply:
Wait if there are 10 steps of 20% size, then we do 200% of the work? XD